### PR TITLE
[Provider] Added Support for MonsterAPI Integration

### PR DESCRIPTION
**Pull request by DheerajQblocks**
**Description:**
This PR adds support for the MonsterAPI provider to our project. It includes the necessary configurations, request/response structures, and transformation logic to integrate MonsterAPI into our project.

**Changes:**
- Added MonsterAPI configuration to `provider/index.ts`.
- Defined MonsterAPI request and response structures in `types.ts`.
- Implemented request transformation logic in `provider/transform.ts`.
- Implemented response transformation logic in `provider/transform.ts`.

**Additional Information:**
- Documentation for MonsterAPI can be found at [MonsterAPI Documentation](https://monsterapi.ai/docs).
- The MonsterAPI API endpoint is `https://llm.monsterapi.ai/v1/generate`.

**Testing:** 
- Tested the integration with MonsterAPI using sample requests.
- Verified the correctness of response transformation logic.

**Screenshots:** 
image

**Related Issue:** #279 (Add MonsterAPI provider)

### CODE REVIEW

**Comment by VisargD** [comment]
Hey @DheerajQblocks . Thanks for the PR! Can you please update this with latest main branch changes. Its missing a lot of major changes that were pushed last week. And the integration does not seem to follow OpenAI transform convention that we follow for other providers. All the response are still in monster API format which is not correct. Please let me know if I am missing something or if you need anymore details on this.

**Comment by DheerajQblocks** [comment]
Hi @VisargD ,

Thank you for reviewing the PR. I appreciate your feedback.

I have updated the PR with the latest changes from the main branch and ensured that it aligns with the OpenAI transform convention as per our standards for other providers. The integration now follows the required format for responses, adhering to the guidelines.

Regarding the response format, I've revised the implementation to conform to the expected structure for MonsterAPI, ensuring that it meets the requirements outlined in the project specifications.

Please review the changes and let me know if there are any further adjustments needed. I'm here to address any additional concerns or questions you may have.

Thank you for your attention to this matter.

Best regards,
@DheerajQblocks

File: package-lock.json
Comment (VisargD) [inline package-lock.json:1]: Can you please remove these changes to package-lock.json? Seems to be unrelated to the PR and are also causing npm install issues.

Comment (DheerajQblocks) [reply package-lock.json:1]: Done

**Comment by DheerajQblocks** [comment]
@VisargD Please review / Approve this PR.

**Comment by VisargD** [comment]
@Dheeraj-Bhandari - Is this correct doc for the API that you have integrated?

**Comment by VisargD** [comment]
@Dheeraj-Bhandari  - Is this correct doc for the API that you have integrated? https://developer.monsterapi.ai/reference/generate_v1_generate_post

**Comment by DheerajQblocks** [comment]
> Is

Yes it's correct

**Comment by DheerajQblocks** [comment]
@VisargD Please Take a Look

**Comment by VisargD** [comment]
Hey @DheerajQblocks - Sorry for the delay here. I will start the review today in sometime.

File: src/providers/monsterapi/index.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import {
+  MonsterAPIChatCompleteConfig,
+  MonsterAPIChatCompleteResponseTransform,
+} from './generate';
+
+const MonsterAPIConfig: ProviderConfigs = {
+  api: MonsterAPIApiConfig,
+  generate: MonsterAPIChatCompleteConfig,
+  responseTransforms: {
+    generate: MonsterAPIChatCompleteResponseTransform,
```
Comment (VisargD) [inline src/providers/monsterapi/index.ts]: Instead of `generate`, the name should be `chatComplete` for this.

File: src/providers/monsterapi/generate.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        code: response.error.code.toString(),
+      },
+      MONSTERAPI
+    );
+  }
+
+  if ('response' in response) {
+    return {
+      id: '',
+      object: '',
```
Comment (VisargD) [inline src/providers/monsterapi/generate.ts]: The `object` value can be `chat.completion` as we are following OpenAI specs

File: src/providers/monsterapi/generate.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      created: 0,
```
Comment (VisargD) [inline src/providers/monsterapi/generate.ts]: For `created`, you can use `Math.floor(Date.now() / 1000)` as value.

File: src/providers/monsterapi/generate.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        param: null,
+        code: response.error.code.toString(),
+      },
+      MONSTERAPI
+    );
+  }
+
+  if ('response' in response) {
+    return {
+      id: '',
```
Comment (VisargD) [inline src/providers/monsterapi/generate.ts]: To have the same response structure across, you can generate a random id and send it in response so that none of the existing SDKs break. We currently use `Date.now().toString()` in some providers as response id.

File: src/providers/monsterapi/generate.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  ChatCompletionResponse,
+  ErrorResponse,
+  ProviderConfig,
+} from '../types';
+import {
+  generateErrorResponse,
+  generateInvalidProviderResponseError,
+} from '../utils';
+
+export const MonsterAPIChatCompleteConfig: ProviderConfig = {
```
Comment (VisargD) [inline src/providers/monsterapi/generate.ts]: As per the API docs, some of the mappings are missing here. Can you please check this once?
If any param is similar to an existing OpenAI param, you can use the OpenAI param name as input and map it to Monster's param. If there is no OpenAI compatible param, then you can add them as a new mapping as it is.

File: src/providers/monsterapi/generate.ts
Comment (VisargD) [inline src/providers/monsterapi/generate.ts:1]: As a part of naming convention that we follow in gateway, please use chatComplete instead of generate.

**Comment by VisargD** [comment]
Hey @Dheeraj-Bhandari - Just checking up on this. Please let me know if there any blockers.

**Comment by DheerajQblocks** [comment]
> Hey @Dheeraj-Bhandari - Just checking up on this. Please let me know if there any blockers.

Sure I'm on with the required changes I will update you once it done

**Comment by DheerajQblocks** [comment]
@VisargD Added the Required changes as per the suggestion provided by you.

Please review/merge this PR ASAP.

File: src/providers/monsterapi/api.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { ProviderAPIConfig } from '../types';
+
+const MonsterAPIApiConfig: ProviderAPIConfig = {
+  getBaseURL: () => 'https://llm.monsterapi.ai/v1',
+  headers: ({ providerOptions }) => {
+    return { Authorization: `Bearer ${providerOptions.apiKey}` };
+  },
+  getEndpoint: ({ fn }) => {
+    switch (fn) {
+      case 'generate':
```
Comment (VisargD) [inline src/providers/monsterapi/api.ts]: This case should also use chatComplete instead of generate. And the default should return an empty string. Basically this:
```
switch (fn) {
      case 'chatComplete':
        return '/generate';
      default:
        return '';
    }
```

File: src/providers/monsterapi/chatComplete.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    param: 'top_k',
+    min: 1,
+    max: 20,
+  },
+  top_p: {
+    param: 'top_p',
+    min: 0,
+    max: 1,
+  },
+  temp: {
```
Comment (VisargD) [inline src/providers/monsterapi/chatComplete.ts]: The openAI compatible name is temperature and not temp. So the input will be temperature which can then be mapped to MonsterAPI specific param name

File: src/providers/monsterapi/index.ts
Comment (VisargD) [inline src/providers/monsterapi/index.ts:1]: We can remove the comments from the file as they are not required and they are a part of github PR review so they should not be a part of the actual code.

Comment (DheerajQblocks) [reply src/providers/monsterapi/index.ts:1]: Done

File: src/providers/monsterapi/api.ts (hunk 6 above)
Comment (DheerajQblocks) [reply src/providers/monsterapi/api.ts]: Done

File: src/providers/monsterapi/chatComplete.ts (hunk 7 above)
Comment (DheerajQblocks) [reply src/providers/monsterapi/chatComplete.ts]: Done

**Comment by DheerajQblocks** [comment]
@VisargD Required changes made. Please review PR

**Comment by VisargD** [comment]
Looks good to me. There is one problem that I found. I am not sure if its expected. The API error response structure that you have used in the chatComplete transformer is not aligned with monster API docs. The docs shows 2 types of error response:

```
{
  "detail": "\n    \"Invalid input payload, could be following.1 The specified model was not found, \"\n    1. Invalid Model\n    2. Bad payload\n"
}
```

```
{
  "detail": [
    {
      "loc": [
        "string",
        0
      ],
      "msg": "string",
      "type": "string"
    }
  ]
}
```

But the code that is written in this PR does not handle any of the above.

**Comment by VisargD** [comment]
Hey @DheerajQblocks  - Can you please check the last comment and the merge conflicts. Everything else looks good to me.

**Comment by DheerajQblocks** [comment]
> Hey @DheerajQblocks  - Can you please check the last comment and the merge conflicts. Everything else looks good to me.

Sure

**Comment by DheerajQblocks** [comment]
> Looks good to me. There is one problem that I found. I am not sure if its expected. The API error response structure that you have used in the chatComplete transformer is not aligned with monster API docs. The docs shows 2 types of error response:
> 
> ```
> {
>   "detail": "\n    \"Invalid input payload, could be following.1 The specified model was not found, \"\n    1. Invalid Model\n    2. Bad payload\n"
> }
> ```
> 
> ```
> {
>   "detail": [
>     {
>       "loc": [
>         "string",
>         0
>       ],
>       "msg": "string",
>       "type": "string"
>     }
>   ]
> }
> ```
> 
> But the code that is written in this PR does not handle any of the above.

Hi @VisargD 
I think our docs are not updated so you can ignore this.
Although I am sending the exact same error response which I am getting from API.

**Comment by DheerajQblocks** [comment]
> Hey @DheerajQblocks - Can you please check the last comment and the merge conflicts. Everything else looks good to me.

@VisargD I have resolved the conflict. Please Review / Merge it on periority.

**Review by VisargD — APPROVED** [review]
